### PR TITLE
shutdown: get only active md arrays.

### PR DESCRIPTION
**Pull request by mtkaczyk**
Hi all,
It is a fix for md cleanup functionality in systemd-shutdown, described in #17106.
> 1. Partitions upon md device should be ignored. They are not mddevs so they simply ignore ioctl.
> 2. mdadm is tending to drop ioctl in flavor of sysfs, so better will be to write "inactive" to md/array_state for array but it is not serious issue.
> 3. For external metadata management, like IMSM is, this routine stops container (md127), but it cannot stop container member > volume (md126) because it is mounted as '/'. So, the simplest way is to ignore containers from stopping.
> They will be stopped later.
> There is a property:
> MD_LEVEL=container
> and in can be used to distinguish containers from arrays.

1. Done.
2. I decided to drop it and stay with ioctl.
3. Done.

Issue mentioned in #17106 is fixed.

Previous md shutdown trace:
```
systemd-shutdown[1]: Stopping MD devices.
systemd-shutdown[1]: Stopping MD /dev/md127 (9:127).
md: md127 stopped.
systemd-shutdown[1]: Stopping MD /dev/md126p5 (259:8).
systemd-shutdown[1]: Could not stop MD /dev/md126p5: Device or resource busy
systemd-shutdown[1]: Stopping MD /dev/md126p3 (259:6).
systemd-shutdown[1]: Could not stop MD /dev/md126p3: Device or resource busy
systemd-shutdown[1]: Stopping MD /dev/md126p2 (259:5).
systemd-shutdown[1]: Could not stop MD /dev/md126p2: Device or resource busy
systemd-shutdown[1]: Stopping MD /dev/md126p1 (259:4).
systemd-shutdown[1]: Could not stop MD /dev/md126p1: Device or resource busy
systemd-shutdown[1]: Stopping MD /dev/md126 (9:126).
systemd-shutdown[1]: Could not stop MD /dev/md126: Device or resource busy
systemd-shutdown[1]: Not all MD devices stopped, 6 left.
systemd-shutdown[1]: Cannot finalize remaining MD devices, continuing.
systemd-shutdown[1]: Successfully changed into root pivot.
systemd-shutdown[1]: Returning to initrd...
```
Trace with my patch applied:
```
systemd-shutdown[1]: Stopping MD devices.
systemd-shutdown[1]: Stopping MD /dev/md126 (9:126).
systemd-shutdown[1]: Could not stop MD /dev/md126: Device or resource busy
systemd-shutdown[1]: Cannot finalize remaining MD devices, continuing.
systemd-shutdown[1]: Successfully changed into root pivot.
systemd-shutdown[1]: Returning to initrd...
```

### CODE REVIEW

File: src/shutdown/umount.c [hunk 1]
```diff
@@ -352,16 +352,30 @@ static int md_list_get(MountPoint **head) {
...
                 MountPoint *m;
                 dev_t devnum;
 
                 if (sd_device_get_devnum(d, &devnum) < 0 ||
                     sd_device_get_devname(d, &dn) < 0)
                         continue;
 
+                r = sd_device_get_property_value(d, "MD_LEVEL", &md_level);
+                if (r < 0) {
+                        log_error("Failed to get MD_LEVEL property for %s, ignoring", dn);
```
Comment (poettering) [inline src/shutdown/umount.c]: errors we ignore and proceed should be logged at most at log_warning level.

also, please always log the error you get. i.e. use:

```c
log_warning_errno(r, "Failed to …, ignoring: %m", dn);
```

File: src/shutdown/umount.c [hunk 2]
```diff
@@ -352,16 +352,30 @@ static int md_list_get(MountPoint **head) {
...
                 return r;
 
+        r = sd_device_enumerator_add_match_property(e, "DEVTYPE", "disk");
+        if (r < 0)
+                return r;
+
         FOREACH_DEVICE(e, d) {
                 _cleanup_free_ char *p = NULL;
                 const char *dn;
+                const char *md_level = NULL;
```
Comment (poettering) [inline src/shutdown/umount.c]: please merge with previous line

File: src/shutdown/umount.c [hunk 3]
```diff
@@ -352,16 +352,30 @@ static int md_list_get(MountPoint **head) {
...
                     sd_device_get_devname(d, &dn) < 0)
                         continue;
 
+                r = sd_device_get_property_value(d, "MD_LEVEL", &md_level);
+                if (r < 0) {
+                        log_error("Failed to get MD_LEVEL property for %s, ignoring", dn);
+                        continue;
+                }
+
+                if (streq(md_level, "container"))
```
Comment (poettering) [inline src/shutdown/umount.c:378]: a comment would be good explaining what "container" means

File: src/shutdown/umount.c [hunk 4]
```diff
@@ -352,16 +352,30 @@ static int md_list_get(MountPoint **head) {
         if (r < 0)
                 return r;
 
+        r = sd_device_enumerator_add_match_property(e, "DEVTYPE", "disk");
```
Comment (poettering) [inline src/shutdown/umount.c:356]: a comment would be good explaining what this is supposed to ensure, i.e. filter out partitions

**Comment by mtkaczyk** [comment]
Hi, @poettering 
New version submitted.

Thanks,
Mariusz

File: src/shutdown/umount.c [hunk 5]
```diff
@@ -347,21 +347,36 @@ static int md_list_get(MountPoint **head) {
...
                 dev_t devnum;
 
                 if (sd_device_get_devnum(d, &devnum) < 0 ||
                     sd_device_get_devname(d, &dn) < 0)
                         continue;
 
+                r = sd_device_get_property_value(d, "MD_LEVEL", &md_level);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to get MD_LEVEL property for %s, ignoring: %m", dn);
+                        continue;
```
Comment (poettering) [inline src/shutdown/umount.c:373]: where did the closing `}` go?

File: src/shutdown/umount.c [hunk 6]
```diff
@@ -347,21 +347,36 @@ static int md_list_get(MountPoint **head) {
...
+
+                /* Container is special type of MD device, used by external metadata.
+                 * One or more active arrays could have reference to the one container.
+                 * It doesn't provide RAID functionality itself, it don't need to be stopped.*/
```
Comment (poettering) [inline src/shutdown/umount.c]: Maybe reword:

```c
/* MD "containers" are a special type of MD devices, used for external metadata. Since it doesn't provide RAID functionality in itself we don't need to stop it. */
```

**Comment by yuwata** [comment]
Not sure which line causes that, but most important CIs are red...

**Comment by mtkaczyk** [comment]
> Not sure which line causes that, but most important CIs are red...
I lost closing '}'. Fixed now.

**Review by yuwata — APPROVED** [review]
Superficially LGTM.

File: src/shutdown/umount.c [hunk 7]
```diff
@@ -347,21 +347,36 @@ static int md_list_get(MountPoint **head) {
         r = sd_device_enumerator_add_match_subsystem(e, "block", true);
         if (r < 0)
                 return r;
-
```
Comment (bluca) [inline src/shutdown/umount.c]: Why was this newline removed?

File: src/shutdown/umount.c [hunk 8]
```diff
@@ -347,21 +347,36 @@ static int md_list_get(MountPoint **head) {
...
                         continue;
 
+                r = sd_device_get_property_value(d, "MD_LEVEL", &md_level);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to get MD_LEVEL property for %s, ignoring: %m", dn);
+                        continue;
+                }
+
+                /* MD "containers" are a special type of MD devices, used for external metadata.
+                * Since it doesn't provide RAID functionality in itself we don't need to stop it. */
```
Comment (bluca) [inline src/shutdown/umount.c]: nit: needs one further space character indentation (`*` should be aligned)

**Comment by poettering** [comment]
lgtm

**Comment by poettering** [comment]
oops, accidental close.

**Comment by bluca** [comment]
Weird, but likely unrelated:

```
Apr 06 17:29:12 H testsuite-70.sh[267]: + tpm2_pcrextend 7:sha256=0000000000000000000000000000000000000000000000000000000000000000
Apr 06 17:29:21 H kernel: tpm tpm0: Operation Timed out
```
